internal/provider/hypercore_vm_resource.go

```diff
@@ -405,3 +415,91 @@ func (r *HypercoreVMResource) ImportState(ctx context.Context, req resource.Impo
 	tflog.Info(ctx, "TTRT HypercoreVMResource IMPORT_STATE")
 	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
 }
+
+func shutdownVM(ctx context.Context, vmUUID string, restClient *utils.RestClient) diag.Diagnostic {
+	currentState, err := utils.GetVMPowerState(vmUUID, *restClient)
+	if err != nil {
+		return err
+	}
+	if currentState == "SHUTOFF" {
+		return nil
+	}
+	// VM needs to be shutdown
+
+	// VM shutdown might be already initiated, but not yet fully done.
+	// Send ACPI shutdown if needed.
+	desiredState, err := utils.GetVMDesiredState(vmUUID, *restClient)
+	if err != nil {
+		return err
+	}
+	if desiredState != "SHUTOFF" {
+		err = utils.ModifyVMPowerState(*restClient, vmUUID, "SHUTDOWN", ctx)
+		if err != nil {
+			return err
+		}
+	}
+	var maxWaitTime time.Duration = 300
+	envMaxWaitTime := os.Getenv("HC_VM_SHUTDOWN_TIMEOUT")
+	if envMaxWaitTime != "" {
+		val, err := strconv.Atoi(envMaxWaitTime)
+		if err != nil {
+			err_msg := fmt.Sprintf("TTRT HypercoreVMResource Destroy, environ HC_VM_SHUTDOWN_TIMEOUT=%s is not number, err=%s", envMaxWaitTime, err)
+			tflog.Error(ctx, err_msg)
+			var diags diag.Diagnostics
+			diags.AddError(
+				"Invalid environ variable HC_VM_SHUTDOWN_TIMEOUT",
+				err_msg,
+			)
+			return diags[0]
+		}
+		maxWaitTime = time.Duration(val)
+	}
+	var waitSleep time.Duration = 5
+	var waitTime time.Duration = 0
+	for {
+		// wait on VM to stop
+		currentState, err := utils.GetVMPowerState(vmUUID, *restClient)
+		if err != nil {
+			return err
+		}
+		if currentState == "SHUTOFF" {
+			return nil
+		}
+		if waitTime > maxWaitTime {
+			break
+		}
+		time.Sleep(waitSleep * time.Second)
```

Comment:
For this linter error: `Multiplication of durations: 'waitSleep * time.Second' (durationcheck)`. Apparently this is a common occurrence in Golang.

According to research, `time.Duration` is just another way of saying `int64`. So a way to fix this could be the following:
```suggestion
	var waitSleep time.Duration = 5 * time.Second  // from ns to s
	var waitTime time.Duration = 0
	for {
		// wait on VM to stop
		currentState, err := utils.GetVMPowerState(vmUUID, *restClient)
		if err != nil {
			return err
		}
		if currentState == "SHUTOFF" {
			return nil
		}
		if waitTime > maxWaitTime {
			break
		}
		time.Sleep(waitSleep)
```